File: Frog.cpp
// Author : ysh
// 2024/08/15 Thu 00:15:21
// https://codeforces.com/group/j50SrY6hsZ/contest/542817/problem/A
#include<bits/stdc++.h>
using namespace std;
const int R = int(1e9) + 7;
#define int long long
array<int,3> operator*(array<int,3>a,array<int,3> b) {
    return {a[0] * b[0] + (a[1] * b[1] % R) * (a[2] - 1),a[1] * b[0] + a[0] * b[1] + (a[1] * b[1] % R) * (a[2] - 2),a[2]};
}
array<int,3> operator%(array<int,3>a,int b) {
    return {a[0] % b,a[1] % b,a[2]};
}
auto check(auto a,int b) {
    if(b == 1) return a;
    auto tmp = check(a,b >> 1);
    tmp = tmp * tmp % R;

    if(b & 1) return tmp * a % R;
    return tmp;
}
inline int rev(int x) {
    return check(x,R - 1 - 1);
}
signed main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int a,b;cin>>a>>b;
    int tmp = rev(a - 1);
    array<int,3> left({0,tmp,a}),right({1,0,a});

    cout<<(check(left,b) * right)[0];
    return 0;
}

File: Frog.py
# Author : ysh
# 2024/08/15 Thu 00:13:58
# https://codeforces.com/group/j50SrY6hsZ/contest/542817/problem/A
R = 10 ** 9 + 7

def check(a: int,b: int) -> int:
    if b == 0: return 1;
    tmp = check(a,b >> 1)
    tmp = tmp * tmp % R

    return tmp * a % R if b & 1 else tmp

def rev(x: int) -> int:
    return check(x,R - 1 - 1)

n,m = [int(i) for i in input().split()]

def lt(a,b):
    return [i % R for i in [a[0] * b[0] + a[1] * b[1] * (n - 1),a[1] * b[0] + a[0] * b[1] + a[1] * b[1] * (n - 2)]]

def ck(a,b: int):
    if b == 1: return a;
    tmp = ck(a,b >> 1)
    tmp = lt(tmp,tmp)

    return lt(tmp,a) if b & 1 else tmp

tmp = rev(n - 1)
left = [0,tmp]
right = [1,0]

print(lt(ck(left,m),right)[0])
quit()